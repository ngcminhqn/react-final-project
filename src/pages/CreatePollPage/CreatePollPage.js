import { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { usersActions } from "../../features/users/usersSlice";
import { Button, Form, Input, notification } from "antd";
import { questionsActions } from "../../features/questions/questionsSlice";
import { EMPTY_STRING } from "../../constants/common";
import {
  CreatePollContainerStyled,
  FormStyled,
  SubTitleStyled,
  TitleStyled,
} from "./CreatePollPage.styled";
import { getUser } from "../../helpers/user";
import Loading from "../../components/Loading/Loading";
import { useNavigate } from "react-router-dom";
import { PATH_NAME } from "../../constants/pathName";

const CreatePollPage = () => {
  const dispatch = useDispatch();
  const [form] = Form.useForm();
  const navigate = useNavigate();

  const { loading } = useSelector((state) => state.questions);
  const user = getUser();

  const openNotification = () => {
    notification.open({
      message: "Success!",
      description: "Create new poll successfully",
    });
  };

  useEffect(() => {
    dispatch(usersActions.handleGetListUsers());
  }, [dispatch]);

  const onFinish = (values) => {
    const payload = {
      author: user?.name || EMPTY_STRING,
      firstOption: values.firstOption,
      secondOption: values.secondOption,
    };
    dispatch(
      questionsActions.handleSaveNewQuestion({
        data: payload,
        onSuccess: () => {
          form.resetFields();
          openNotification();
          navigate(PATH_NAME.HOME);
        },
      })
    );
  };

  return (
    <CreatePollContainerStyled>
      <Loading isLoading={loading} />
      <TitleStyled>Would You Rather</TitleStyled>
      <SubTitleStyled>Create Your Own Poll</SubTitleStyled>
      <FormStyled
        form={form}
        name="create_poll_form"
        onFinish={onFinish}
        layout="vertical"
      >
        <Form.Item
          label="First Option"
          required
          name="firstOption"
          rules={[{ required: true, message: "Please input first option" }]}
        >
          <Input placeholder="Option One" />
        </Form.Item>
        <Form.Item
          label="Second Option"
          required
          name="secondOption"
          rules={[{ required: true, message: "Please input second option" }]}
        >
          <Input placeholder="Option Two" />
        </Form.Item>

        <Form.Item>
          <Button type="primary" htmlType="submit">
            Submit
          </Button>
        </Form.Item>
      </FormStyled>
    </CreatePollContainerStyled>
  );
};

export default CreatePollPage;
