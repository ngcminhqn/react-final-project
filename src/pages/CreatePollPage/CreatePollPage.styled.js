import { Form } from "antd";
import styled from "styled-components";

export const CreatePollContainerStyled = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  height: 100vh;
  padding: 20px 0;

  .ant-form-vertical .ant-form-item-row {
    align-items: center;
  }
`;

export const TitleStyled = styled.p`
  color: black;
  font-weight: bold;
  font-size: 30px;
  text-align: center;
  margin: 10px 0;
`;

export const SubTitleStyled = styled.p`
  color: gray;
  font-weight: bold;
  font-size: 16px;
  text-align: center;
`;

export const FormStyled = styled(Form)`
  width: 80%;

  :last-child {
    display: flex;
    justify-content: center;
  }
`;

export const ErrorMessage = styled.p`
  color: red;
`;
