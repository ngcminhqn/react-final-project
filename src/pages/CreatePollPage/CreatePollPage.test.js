import { renderWithProviders } from "../../utils/test-utils";
import { fireEvent, waitFor } from "@testing-library/react";
import CreatePollPage from "./CreatePollPage";

describe("CreatePollPage", () => {
  it("render CreatePollPage", async () => {
    const { container } = renderWithProviders(<CreatePollPage />);
    expect(container).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  it("validation form", async () => {
    const { getByRole, getByText } = renderWithProviders(<CreatePollPage />);
    const submitButton = getByRole("button", { name: /Submit/i });
    fireEvent.submit(submitButton);
    await waitFor(() => {
      const userError = getByText(/Please input first option/i);
      const passwordError = getByText(/Please input first option/i);
      expect(userError).toBeTruthy();
      expect(passwordError).toBeTruthy();
    });
  });

  it("show errormessage when the option 1 field is empty", async () => {
    const { getByRole, findByText, getByPlaceholderText } = renderWithProviders(
      <CreatePollPage />
    );

    const opt1 = getByPlaceholderText(/Option One/i);
    const submitButton = getByRole("button", { name: /Submit/i });

    fireEvent.change(opt1, { target: { value: "optionone" } });
    fireEvent.submit(submitButton);

    const errorMessage = await findByText("Please input second option");
    expect(errorMessage).toBeTruthy();
  });

  it("show errormessage when the option 2 field is empty", async () => {
    const { getByRole, findByText, getByPlaceholderText } = renderWithProviders(
      <CreatePollPage />
    );

    const opt2 = getByPlaceholderText(/Option Two/i);
    const submitButton = getByRole("button", { name: /Submit/i });

    fireEvent.change(opt2, { target: { value: "optiontwo" } });
    fireEvent.submit(submitButton);

    const errorMessage = await findByText("Please input first option");
    expect(errorMessage).toBeTruthy();
  });
});
