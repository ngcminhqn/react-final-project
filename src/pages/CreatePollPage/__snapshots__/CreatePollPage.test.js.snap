// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePollPage render CreatePollPage 1`] = `
<div>
  <div
    class="sc-aYaIB fseERr"
  >
    <p
      class="sc-gEvDqW enkQJm"
    >
      Would You Rather
    </p>
    <p
      class="sc-eqUzNf fgdwwx"
    >
      Create Your Own Poll
    </p>
    <form
      class="ant-form ant-form-vertical css-dev-only-do-not-override-m4timi sc-fqkwJk buqAUq"
      id="create_poll_form"
    >
      <div
        class="ant-form-item css-dev-only-do-not-override-m4timi"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-m4timi"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-m4timi"
          >
            <label
              class="ant-form-item-required"
              for="create_poll_form_firstOption"
              title="First Option"
            >
              First Option
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-m4timi"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input css-dev-only-do-not-override-m4timi ant-input-outlined"
                  id="create_poll_form_firstOption"
                  placeholder="Option One"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-m4timi"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-m4timi"
        >
          <div
            class="ant-col ant-form-item-label css-dev-only-do-not-override-m4timi"
          >
            <label
              class="ant-form-item-required"
              for="create_poll_form_secondOption"
              title="Second Option"
            >
              Second Option
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-m4timi"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  aria-required="true"
                  class="ant-input css-dev-only-do-not-override-m4timi ant-input-outlined"
                  id="create_poll_form_secondOption"
                  placeholder="Option Two"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-form-item css-dev-only-do-not-override-m4timi"
      >
        <div
          class="ant-row ant-form-item-row css-dev-only-do-not-override-m4timi"
        >
          <div
            class="ant-col ant-form-item-control css-dev-only-do-not-override-m4timi"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn css-dev-only-do-not-override-m4timi ant-btn-primary"
                  type="submit"
                >
                  <span>
                    Submit
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
